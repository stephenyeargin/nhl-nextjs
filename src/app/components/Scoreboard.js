import React from 'react';

import { PERIOD_DESCRIPTORS } from '../utils/constants';
import { PropTypes } from 'prop-types';
import TeamLogo from './TeamLogo';

const Scoreboard = ({ game, linescore }) => {
  // Assuming game has a total period count, if not, adjust accordingly
  const totalPeriods = game.periodDescriptor?.number > 3 ? game.periodDescriptor.number : 3;

  return (
    <table className="w-full text-sm">
      <thead>
        <tr>
          <th className="w-3/12 border bg-slate-200 dark:bg-slate-800 p-2"></th>
          {Array.from({ length: totalPeriods }).map((_, index) => {
            const periodNumber = index + 1;
            const period = linescore.byPeriod.find(p => p.periodDescriptor.number === periodNumber);
            
            return (
              <th
                className="w-1/12 text-center border bg-slate-200 dark:bg-slate-800 p-2"
                key={periodNumber}
              >
                {period ? (
                  period.periodDescriptor.periodType === 'REG' ? (
                    <>{PERIOD_DESCRIPTORS[period.periodDescriptor.number]}</>
                  ) : (
                    <>{period.periodDescriptor.periodType}</>
                  )
                ) : (
                  <>{PERIOD_DESCRIPTORS[index + 1]}</>
                )}
              </th>
            );
          })}
          <th className="w-1/12 text-center border bg-slate-200 dark:bg-slate-800 p-2">T</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td className="p-2 text-center border p-2">
            <div className="flex items-center">
              <TeamLogo
                src={game.awayTeam.logo}
                className="w-10 mr-1"
                alt="Logo"
              />
              <span className="font-bold">{game.awayTeam.abbrev}</span>
            </div>
          </td>
          {Array.from({ length: totalPeriods }).map((_, index) => {
            const periodNumber = index + 1;
            const period = linescore.byPeriod.find(p => p.periodDescriptor.number === periodNumber);
            
            return (
              <td className="text-center border p-2" key={periodNumber}>
                {period ? period.away : '–'}
              </td>
            );
          })}
          <td className="text-center border p-2 font-bold">
            {linescore.totals.away}
          </td>
        </tr>
        <tr>
          <td className="p-2 text-center border p-2">
            <div className="flex items-center">
              <TeamLogo
                src={game.homeTeam.logo}
                className="w-10 mr-1"
                alt="Logo"
              />
              <span className="font-bold">{game.homeTeam.abbrev}</span>
            </div>
          </td>
          {Array.from({ length: totalPeriods }).map((_, index) => {
            const periodNumber = index + 1;
            const period = linescore.byPeriod.find(p => p.periodDescriptor.number === periodNumber);
            
            return (
              <td className="text-center border" key={periodNumber}>
                {period ? period.home : '–'}
              </td>
            );
          })}
          <td className="text-center border p-2 font-bold">
            {linescore.totals.home}
          </td>
        </tr>
      </tbody>
    </table>
  );
};

Scoreboard.propTypes = {
  game: PropTypes.object.isRequired,
  linescore: PropTypes.object.isRequired,
};

export default Scoreboard;
