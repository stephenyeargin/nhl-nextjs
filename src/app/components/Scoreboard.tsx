import React from 'react';
import TeamLogo from './TeamLogo';
import { formatPeriodLabel } from '../utils/formatters';
import type { PeriodDescriptor } from '@/app/types/game';

interface TeamInfoLite {
  logo?: string;
  abbrev: string;
}

interface ScoreboardPeriodDescriptor extends PeriodDescriptor {
  number: number;
}

interface GameForScoreboard {
  periodDescriptor: ScoreboardPeriodDescriptor;
  regPeriods: number;
  awayTeam: TeamInfoLite;
  homeTeam: TeamInfoLite;
}

interface LineScorePeriod {
  periodDescriptor: { number: number };
  away: number;
  home: number;
}

interface LineScoreTotals {
  away: number;
  home: number;
}

interface LineScore {
  byPeriod: LineScorePeriod[];
  totals: LineScoreTotals;
}

interface ScoreboardProps {
  game: GameForScoreboard;
  linescore: LineScore;
}

const Scoreboard: React.FC<ScoreboardProps> = ({ game, linescore }) => {
  const totalPeriods =
    game.periodDescriptor?.number < game.regPeriods
      ? game.regPeriods
      : game.periodDescriptor?.number;

  // Helper function to determine if a period is an overtime period with no scores
  const isEmptyOvertime = (period: LineScorePeriod) => {
    return (
      period.periodDescriptor.number > game.regPeriods && // Check if it's overtime
      period.away === 0 &&
      period.home === 0 // Both teams did not score
    );
  };

  return (
    <table className="w-full text-sm">
      <thead>
        <tr className="border">
          <th className="w-3/12 bg-slate-200 dark:bg-slate-800 p-2"></th>
          {Array.from({ length: totalPeriods }).map((_, index) => {
            const periodNumber = index + 1;
            const period = linescore.byPeriod.find(
              (p) => p.periodDescriptor.number === periodNumber
            );

            // Skip rendering period column for empty overtime periods
            if (period && isEmptyOvertime(period)) {
              return null;
            }

            return (
              <th
                className="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
                key={`period-${periodNumber}`}
              >
                {formatPeriodLabel({ ...game.periodDescriptor, number: periodNumber })}
              </th>
            );
          })}
          <th className="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2">T</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td className="p-2 text-center border">
            <div className="flex flex-wrap items-center gap-1">
              <TeamLogo
                src={game.awayTeam.logo}
                team={game.awayTeam.abbrev}
                className="w-10"
                alt="Logo"
              />
              <div className="font-bold">{game.awayTeam.abbrev}</div>
            </div>
          </td>
          {Array.from({ length: totalPeriods }).map((_, index) => {
            const periodNumber = index + 1;
            const period = linescore.byPeriod.find(
              (p) => p.periodDescriptor.number === periodNumber
            );

            // Skip rendering the score for empty overtime periods
            if (period && isEmptyOvertime(period)) {
              return null;
            }

            return (
              <td className="text-center border p-2" key={periodNumber}>
                {period ? period.away : '-'}
              </td>
            );
          })}
          <td className="text-center border p-2 font-bold">{linescore.totals.away}</td>
        </tr>
        <tr>
          <td className="p-2 text-center border">
            <div className="flex flex-wrap items-center gap-1">
              <TeamLogo
                src={game.homeTeam.logo}
                team={game.homeTeam.abbrev}
                className="w-10"
                alt="Logo"
              />
              <div className="font-bold">{game.homeTeam.abbrev}</div>
            </div>
          </td>
          {Array.from({ length: totalPeriods }).map((_, index) => {
            const periodNumber = index + 1;
            const period = linescore.byPeriod.find(
              (p) => p.periodDescriptor.number === periodNumber
            );

            // Skip rendering the score for empty overtime periods
            if (period && isEmptyOvertime(period)) {
              return null;
            }

            return (
              <td className="text-center border" key={periodNumber}>
                {period ? period.home : '-'}
              </td>
            );
          })}
          <td className="text-center border p-2 font-bold">{linescore.totals.home}</td>
        </tr>
      </tbody>
    </table>
  );
};

export default Scoreboard;
