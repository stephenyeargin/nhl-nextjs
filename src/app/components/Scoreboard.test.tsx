import React from 'react';
import { render, screen } from '@testing-library/react';
import Scoreboard from './Scoreboard';

jest.mock('./TeamLogo', () => {
  const MockTeamLogo = ({ team }: any) => <div data-testid={`logo-${team}`} />;
  MockTeamLogo.displayName = 'MockTeamLogo';

  return MockTeamLogo;
});
jest.mock('../utils/formatters', () => ({
  formatPeriodLabel: ({ number }: any) => `${number}`,
}));

const makeGame = (overrides: any = {}) => ({
  periodDescriptor: { number: 5, periodType: 'OT' },
  regPeriods: 3,
  awayTeam: { abbrev: 'AWY' },
  homeTeam: { abbrev: 'HOM' },
  ...overrides,
});

const makeLinescore = (periods: any[], totals = { away: 2, home: 3 }) => ({
  byPeriod: periods,
  totals,
});

describe('Scoreboard', () => {
  test('renders periods skipping empty overtime periods', () => {
    const linescore = makeLinescore([
      { periodDescriptor: { number: 1 }, away: 1, home: 0 },
      { periodDescriptor: { number: 2 }, away: 0, home: 1 },
      { periodDescriptor: { number: 3 }, away: 1, home: 1 },
      { periodDescriptor: { number: 4 }, away: 0, home: 0 }, // empty OT (skipped)
      { periodDescriptor: { number: 5 }, away: 0, home: 0 }, // empty 2OT (skipped)
    ]);
    render(<Scoreboard game={makeGame()} linescore={linescore} />);
    // Column headers only include 1,2,3 and T
    const headers = screen.getAllByRole('columnheader').map(h => (h as HTMLElement).textContent);
    expect(headers).toEqual(['', '1', '2', '3', 'T']);
    // Ensure skipped OT headers 4 and 5 are absent
    expect(headers.includes('4')).toBe(false);
    expect(headers.includes('5')).toBe(false);
  });
});
