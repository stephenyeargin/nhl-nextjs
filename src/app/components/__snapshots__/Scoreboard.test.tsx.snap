// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard renders overtime period when it has scoring 1`] = `
<DocumentFragment>
  <table
    class="w-full text-sm"
  >
    <thead>
      <tr
        class="border"
      >
        <th
          class="w-3/12 bg-slate-200 dark:bg-slate-800 p-2"
        />
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          1
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          2
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          3
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          4
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          T
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="p-2 text-center border"
        >
          <div
            class="flex flex-wrap items-center gap-1"
          >
            <div
              data-testid="logo-AWY"
            />
            <div
              class="font-bold"
            >
              AWY
            </div>
          </div>
        </td>
        <td
          class="text-center border p-2"
        >
          1
        </td>
        <td
          class="text-center border p-2"
        >
          0
        </td>
        <td
          class="text-center border p-2"
        >
          1
        </td>
        <td
          class="text-center border p-2"
        >
          1
        </td>
        <td
          class="text-center border p-2 font-bold"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          class="p-2 text-center border"
        >
          <div
            class="flex flex-wrap items-center gap-1"
          >
            <div
              data-testid="logo-HOM"
            />
            <div
              class="font-bold"
            >
              HOM
            </div>
          </div>
        </td>
        <td
          class="text-center border"
        >
          0
        </td>
        <td
          class="text-center border"
        >
          1
        </td>
        <td
          class="text-center border"
        >
          1
        </td>
        <td
          class="text-center border"
        >
          0
        </td>
        <td
          class="text-center border p-2 font-bold"
        >
          2
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Scoreboard renders periods skipping empty overtime periods 1`] = `
<DocumentFragment>
  <table
    class="w-full text-sm"
  >
    <thead>
      <tr
        class="border"
      >
        <th
          class="w-3/12 bg-slate-200 dark:bg-slate-800 p-2"
        />
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          1
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          2
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          3
        </th>
        <th
          class="w-1/12 text-center bg-slate-200 dark:bg-slate-800 p-2"
        >
          T
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="p-2 text-center border"
        >
          <div
            class="flex flex-wrap items-center gap-1"
          >
            <div
              data-testid="logo-AWY"
            />
            <div
              class="font-bold"
            >
              AWY
            </div>
          </div>
        </td>
        <td
          class="text-center border p-2"
        >
          1
        </td>
        <td
          class="text-center border p-2"
        >
          0
        </td>
        <td
          class="text-center border p-2"
        >
          1
        </td>
        <td
          class="text-center border p-2 font-bold"
        >
          2
        </td>
      </tr>
      <tr>
        <td
          class="p-2 text-center border"
        >
          <div
            class="flex flex-wrap items-center gap-1"
          >
            <div
              data-testid="logo-HOM"
            />
            <div
              class="font-bold"
            >
              HOM
            </div>
          </div>
        </td>
        <td
          class="text-center border"
        >
          0
        </td>
        <td
          class="text-center border"
        >
          1
        </td>
        <td
          class="text-center border"
        >
          1
        </td>
        <td
          class="text-center border p-2 font-bold"
        >
          3
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
